homeassistant/components/totalconnect/config_flow.py

```diff
@@ -44,13 +57,96 @@ async def async_step_user(self, user_input=None):
+                if self.usercodes[location_id] is None:
+                    complete = False
+
+            if not errors and complete:
+                return self.async_create_entry(
+                    title="Total Connect",
+                    data={
+                        CONF_USERNAME: self.username,
+                        CONF_PASSWORD: self.password,
+                        CONF_USERCODES: self.usercodes,
+                    },
+                )
+        else:
+            for location_id in self.client.locations:
+                self.usercodes[location_id] = None
+
+        # show the next location that needs a usercode
+        location_codes = {}
+        for location_id in self.usercodes:
+            if self.usercodes[location_id] is None:
+                location_codes[
+                    vol.Required(
+                        CONF_LOCATION,
+                        default=location_id,
+                    )
+                ] = str
+                break
+
+        data_schema = vol.Schema(location_codes)
+        return self.async_show_form(
+            step_id="locations",
+            data_schema=data_schema,
+            errors=errors,
+            description_placeholders={"base": "description"},
+        )
+
+    async def async_step_reauth(self, config):
+        """Perform reauth upon an authentication error or no usercode."""
+        self.username = config[CONF_USERNAME]
+        self.usercodes = config[CONF_USERCODES]
+
+        return await self.async_step_reauth_confirm()
+
+    async def async_step_reauth_confirm(self, user_input=None):
+        """Dialog that informs the user that reauth is required."""
+        errors = {}
+        if user_input is None:
+            return self.async_show_form(
+                step_id="reauth_confirm",
+                data_schema=PASSWORD_DATA_SCHEMA,
+            )
 
-    async def is_valid(self, username="", password=""):
-        """Return true if the given username and password are valid."""
         client = await self.hass.async_add_executor_job(
-            TotalConnectClient.TotalConnectClient, username, password
+            TotalConnectClient.TotalConnectClient,
+            self.username,
+            user_input[CONF_PASSWORD],
+            self.usercodes,
         )
-        return client.is_valid_credentials()
+
+        if not client.is_valid_credentials():
+            errors["base"] = "invalid_auth"
+            return self.async_show_form(
+                step_id="reauth_confirm",
+                errors=errors,
+                data_schema=PASSWORD_DATA_SCHEMA,
+            )
+
+        existing_entry = await self.async_set_unique_id(self.username)
+        new_entry = {
+            CONF_USERNAME: self.username,
+            CONF_PASSWORD: user_input[CONF_PASSWORD],
+            CONF_USERCODES: self.usercodes,
+        }
+        self.hass.config_entries.async_update_entry(existing_entry, data=new_entry)
```

Comment:
We need to make sure that the config entry is reloaded after we update the config entry. Either we do that explicitly here or we add an update listener to the config entry when setting up the config entry.

https://developers.home-assistant.io/docs/config_entries_options_flow_handler#signal-updates

Reply:
Good catch, thanks.  Done